### ISSUE TITLE
(Feature request) Faster logsumexp

### ISSUE THREAD

**Issue by itaysason** [scipy/scipy#9300]
Hi, lately I have been working on statistical models, and for that I work a lot in log-space. The issue is that logsumexp is a VERY slow function (takes about 75% of the program time) most likely because it is implemented in python. Do you think it is possible to implement it faster?

**Comment by ilayn** [comment]
Can you add an example how you use it? Are you using check_finite? Do you have infs or 0s in your arrays. Do you supply Numpy arrays or array_likes? We would really need more information. Otherwise the answer is always yes. But somebody has to do it.

**Comment by itaysason** [comment]
I use numpy arrays, about the rest I don't use check infinite as it doesn't give me any problems. my arrays may have -inf in them.

Here is a small example

```
import numpy as np
from scipy.misc import logsumexp


class ExampleClass:
    def __init__(self, num_states, num_emissions):
        self.num_states = num_states
        self.num_emissions = num_emissions
        self.alpha = np.random.random(num_states)
        self.pi = np.random.dirichlet([0.5] * num_states)
        self.emissions = np.random.dirichlet([0.5] * num_emissions, size=num_states)

    def fit(self, data, max_iterations=10):
        for iteration in range(int(max_iterations)):
            params = self.expectation_step(data)
            self.maximization_step(params)
            print(iteration)

    def expectation_step(self, data):
        num_states = self.num_states
        log_alpha = np.log(self.alpha)
        log_1m_alpha = np.log(1 - self.alpha)
        log_pi = np.log(self.pi)
        log_emissions = np.log(self.emissions.T)
        a = np.zeros(num_states)
        b = np.zeros(num_states)
        c = np.zeros(num_states)
        e = np.zeros((self.num_emissions, num_states))

        for seq in data:
            log_fwd = np.empty((len(seq), num_states))
            log_bwd = np.empty((len(seq), num_states))

            log_fwd[0] = log_pi + log_emissions[seq[0]]
            log_bwd[-1] = 0

            for i in range(1, len(seq)):
                tmp = logsumexp(log_fwd[i - 1] + log_1m_alpha)
                log_fwd[i] = log_emissions[seq[i]] + np.logaddexp(log_pi + tmp, log_fwd[i - 1] + log_alpha)

                tmp = logsumexp(log_bwd[-i] + log_emissions[seq[-i]] + log_pi)
                log_bwd[- i - 1] = np.logaddexp(log_1m_alpha + tmp, log_alpha + log_emissions[seq[-i]] + log_bwd[-i])

            curr_prob = logsumexp(log_fwd[-1])
            curr_a = np.log(np.zeros(a.shape))
            curr_b = log_fwd[0] + log_bwd[0]
            curr_c = np.log(np.zeros(a.shape))
            for i in range(len(seq) - 1):
                curr_a = np.logaddexp(log_fwd[i] + log_alpha + log_emissions[seq[i+1]] + log_bwd[i+1], curr_a)
                for j in range(num_states):
                    curr_b = np.logaddexp(log_fwd[i, j] + log_1m_alpha[j] + log_pi + log_emissions[seq[i+1]] + log_bwd[i+1], curr_b)
                    curr_c = np.logaddexp(log_fwd[i] + log_1m_alpha + log_pi[j] + log_emissions[seq[i+1], j] + log_bwd[i+1, j], curr_c)

            curr_a -= curr_prob
            curr_b -= curr_prob
            curr_c -= curr_prob

            a += np.exp(curr_a)
            b += np.exp(curr_b)
            c += np.exp(curr_c)
            pr_xt_qi = log_fwd + log_bwd - curr_prob
            for i in range(len(seq)):
                e[seq[i]] += np.exp(pr_xt_qi[i])

        return a, b, c, e

    def maximization_step(self, params):
        self.alpha = params[0] / (params[0] + params[2])
        self.pi = params[1] / np.sum(params[1])
        self.emissions = params[3] / np.sum(params[3], 0)
        self.emissions = self.emissions.T


n = 12
m = 96
train_data = [np.random.randint(0, 96, 1000) for i in range(10)]
model = ExampleClass(n, m)
model.fit(train_data)

```

here it takes 30% of the running time but in my whole program much more.

**Comment by rgommers** [comment]
Can we just replace these occurrences with `np.logaddexp.reduce` instead? I'm not sure what this comments is about exactly,
```
NumPy has a logaddexp function which is very similar to `logsumexp`, but
only handles two arguments. `logaddexp.reduce` is similar to this
function, but may be less stable.
```
but it may not be relevant here.

**Comment by itaysason** [comment]
Actually this is quite helpful, I did not know the reduce option!
It really is a little bit less stable, I wander why that is.
I ran a small test to see if it improves, here is the code:
`import time
import numpy as np
from scipy.special import logsumexp

for j in range(1, 6):
    q = np.ones(10 ** j)
    tic1 = time.clock()
    for i in range(1000):
        logsumexp(q)
    tic2 = time.clock()
    for i in range(1000):
        np.logaddexp.reduce(q)
    tic3 = time.clock()
    print('{}: logsumexp - {}, np.logaddexp.reduce - {}'.format(j, tic2 - tic1, tic3 - tic2))`

and this is the output:
1: logsumexp - 0.058854221552807645, np.logaddexp.reduce - 0.0017786322523601489
2: logsumexp - 0.06082768086402461, np.logaddexp.reduce - 0.006411300997524821
3: logsumexp - 0.07094326625764977, np.logaddexp.reduce - 0.05172375682087477
4: logsumexp - 0.12272356919059385, np.logaddexp.reduce - 0.4938943050354313
5: logsumexp - 0.9046031105502194, np.logaddexp.reduce - 4.867595736423151

This is quite interesting, in my use case it is pretty helpful as I am running my code on small arrays (<1000).

I still don't think I can use it as it is less stable then the scipy one but thanks anyway I am sure the reduce functions will be helpful for me in the future!

**Comment by andyfaff** [comment]
To make things like this faster requires profiling, otherwise one doesn't know what to tackle. One way of doing this is to use the `line_profiler` package, see below. From a look at the source, and the profiling I don't think that the code here would benefit from cythonisation. Most of the time is spent in `np.exp`, and `np.sum`. These aren't going to run any faster in cython. It's unavoidable that `exp` is expensive to calculate.
If you know of a faster algorithm that would retain accuracy, then perhaps something could be done.

```
Timer unit: 1e-06 s

Total time: 0.083398 s
File: /Users/andrew/miniconda3/envs/dev35/lib/python3.5/site-packages/scipy/special/_logsumexp.py
Function: logsumexp at line 9

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     9                                           def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
    10                                               """Compute the log of the sum of exponentials of input elements.
    11                                           
    12                                               Parameters
    13                                               ----------
    14                                               a : array_like
    15                                                   Input array.
    16                                               axis : None or int or tuple of ints, optional
    17                                                   Axis or axes over which the sum is taken. By default `axis` is None,
    18                                                   and all elements are summed.
    19                                           
    20                                                   .. versionadded:: 0.11.0
    21                                               keepdims : bool, optional
    22                                                   If this is set to True, the axes which are reduced are left in the
    23                                                   result as dimensions with size one. With this option, the result
    24                                                   will broadcast correctly against the original array.
    25                                           
    26                                                   .. versionadded:: 0.15.0
    27                                               b : array-like, optional
    28                                                   Scaling factor for exp(`a`) must be of the same shape as `a` or
    29                                                   broadcastable to `a`. These values may be negative in order to
    30                                                   implement subtraction.
    31                                           
    32                                                   .. versionadded:: 0.12.0
    33                                               return_sign : bool, optional
    34                                                   If this is set to True, the result will be a pair containing sign
    35                                                   information; if False, results that are negative will be returned
    36                                                   as NaN. Default is False (no sign information).
    37                                           
    38                                                   .. versionadded:: 0.16.0
    39                                           
    40                                               Returns
    41                                               -------
    42                                               res : ndarray
    43                                                   The result, ``np.log(np.sum(np.exp(a)))`` calculated in a numerically
    44                                                   more stable way. If `b` is given then ``np.log(np.sum(b*np.exp(a)))``
    45                                                   is returned.
    46                                               sgn : ndarray
    47                                                   If return_sign is True, this will be an array of floating-point
    48                                                   numbers matching res and +1, 0, or -1 depending on the sign
    49                                                   of the result. If False, only one result is returned.
    50                                           
    51                                               See Also
    52                                               --------
    53                                               numpy.logaddexp, numpy.logaddexp2
    54                                           
    55                                               Notes
    56                                               -----
    57                                               NumPy has a logaddexp function which is very similar to `logsumexp`, but
    58                                               only handles two arguments. `logaddexp.reduce` is similar to this
    59                                               function, but may be less stable.
    60                                           
    61                                               Examples
    62                                               --------
    63                                               >>> from scipy.special import logsumexp
    64                                               >>> a = np.arange(10)
    65                                               >>> np.log(np.sum(np.exp(a)))
    66                                               9.4586297444267107
    67                                               >>> logsumexp(a)
    68                                               9.4586297444267107
    69                                           
    70                                               With weights
    71                                           
    72                                               >>> a = np.arange(10)
    73                                               >>> b = np.arange(10, 0, -1)
    74                                               >>> logsumexp(a, b=b)
    75                                               9.9170178533034665
    76                                               >>> np.log(np.sum(b*np.exp(a)))
    77                                               9.9170178533034647
    78                                           
    79                                               Returning a sign flag
    80                                           
    81                                               >>> logsumexp([1,2],b=[1,-1],return_sign=True)
    82                                               (1.5413248546129181, -1.0)
    83                                           
    84                                               Notice that `logsumexp` does not directly support masked arrays. To use it
    85                                               on a masked array, convert the mask into zero weights:
    86                                           
    87                                               >>> a = np.ma.array([np.log(2), 2, np.log(3)],
    88                                               ...                  mask=[False, True, False])
    89                                               >>> b = (~a.mask).astype(int)
    90                                               >>> logsumexp(a.data, b=b), np.log(5)
    91                                               1.6094379124341005, 1.6094379124341005
    92                                           
    93                                               """
    94         1         21.0     21.0      0.0      a = _asarray_validated(a, check_finite=False)
    95         1          1.0      1.0      0.0      if b is not None:
    96                                                   a, b = np.broadcast_arrays(a, b)
    97                                                   if np.any(b == 0):
    98                                                       a = a + 0.  # promote to at least float
    99                                                       a[b == 0] = -np.inf
   100                                           
   101         1      10415.0  10415.0     12.5      a_max = np.amax(a, axis=axis, keepdims=True)
   102                                           
   103         1          4.0      4.0      0.0      if a_max.ndim > 0:
   104         1         16.0     16.0      0.0          a_max[~np.isfinite(a_max)] = 0
   105                                               elif not np.isfinite(a_max):
   106                                                   a_max = 0
   107                                           
   108         1          0.0      0.0      0.0      if b is not None:
   109                                                   b = np.asarray(b)
   110                                                   tmp = b * np.exp(a - a_max)
   111                                               else:
   112         1      62430.0  62430.0     74.9          tmp = np.exp(a - a_max)
   113                                           
   114                                               # suppress warnings about log of zero
   115         1         36.0     36.0      0.0      with np.errstate(divide='ignore'):
   116         1      10422.0  10422.0     12.5          s = np.sum(tmp, axis=axis, keepdims=keepdims)
   117         1          2.0      2.0      0.0          if return_sign:
   118                                                       sgn = np.sign(s)
   119                                                       s *= sgn  # /= makes more sense but we need zero -> zero
   120         1         42.0     42.0      0.1          out = np.log(s)
   121                                           
   122         1          1.0      1.0      0.0      if not keepdims:
   123         1          7.0      7.0      0.0          a_max = np.squeeze(a_max, axis=axis)
   124         1          1.0      1.0      0.0      out += a_max
   125                                           
   126         1          0.0      0.0      0.0      if return_sign:
   127                                                   return out, sgn
   128                                               else:
   129         1          0.0      0.0      0.0          return out
```

**Comment by ev-br** [comment]
One only possible thing to check with line_profiler could be to split off the subtraction `a-a_max` in the exponential. In numpy this involves a temporary, which cythonized routine may avoid. (I'm not suggesting you do it, Andrew!)

**Comment by andyfaff** [comment]
@ev-br  that seems to have some merit to it:

```
>>> q = np.ones(10 ** 7)
>>> a_max = np.amax(q)
>>> b = q - a_max
>>> %timeit q - a_max
>>> %timeit np.exp(b)
14.3 ms ± 83.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
15.1 ms ± 144 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by ev-br** [comment]
Got curious myself and did a quick experiment: https://gist.github.com/ev-br/0837b336eb0cbdd626d41caa6216b716

As it often happens, it's easy to beat numpy for small arrays, but as the size grows, overhead becomes insignificant. Here, the change is from x60 to mere 30%.

**Comment by andyfaff** [comment]
The cython implementation of `lsm` isn't doing the checking that the pure Python version is, so I'd expect the small array case to be closer once that was taken into account.

**Comment by andyfaff** [comment]
Moreover, running your notebook on my computer gives an opposing effect for the cythonised code.
```
>>> x = np.random.RandomState(123).uniform(size=10**7)
>>> print(lsm(x) - logsumexp(x))
>>> %timeit lsm(x)
>>> %timeit logsumexp(x)
-5.684341886080802e-14
88.2 ms ± 2.62 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
42.4 ms ± 2.77 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
The cython version is slower for large arrays, no idea why

**Comment by ev-br** [comment]
Sure, this notebook is only a quick test of an upper limit of the speedup with cython --- if used in a specific user application, where it's possible to guarantee the input data format etc.

The fact that lsm is slower on your machine and faster on mine might be due to some compilation switches? I tested it on a bog-standard ubuntu 14.04 with stock gcc only.

**Comment by andyfaff** [comment]
macOS + conda python 3.7 + numpy-mkl + cython 0.29 + clang + ?
